libclamav_rust/src/onenote.rs

```diff
@@ -0,0 +1,355 @@
+/*
+ *  Onenote document parser to extract embedded files.
+ *
+ *  Copyright (C) 2023 Cisco Systems, Inc. and/or its affiliates. All rights reserved.
+ *
+ *  Authors: Micah Snyder
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License version 2 as
+ *  published by the Free Software Foundation.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ *  MA 02110-1301, USA.
+ */
+
+use std::mem;
+use std::panic;
+use std::path::PathBuf;
+
+use log::{debug, error};
+use thiserror::Error;
+
+use onenote_parser;
+
+/// OneDumpError enumerates all possible errors returned by this library.
+#[derive(Error, Debug)]
+pub enum OneDumpError {
+    #[error("Invalid format")]
+    Format,
+
+    #[error("Invalid parameter: {0}")]
+    InvalidParameter(String),
+
+    #[error("Failed to open file: {0}, {1}")]
+    FailedToOpen(PathBuf, String),
+
+    #[error("Failed to get size for file: {0}")]
+    FailedToGetFileSize(PathBuf),
+
+    #[error("{0} parameter is NULL")]
+    NullParam(&'static str),
+
+    #[error("No more files to extract")]
+    NoMoreFiles,
+
+    #[error("Unable to parse OneNote file")]
+    Parse,
+
+    #[error("Failed to load image due to bug in image decoder")]
+    SectionParsePanic(),
+}
+
+fn find_bytes(haystack: &[u8], needle: &[u8]) -> Option<usize> {
+    haystack
+        .windows(needle.len())
+        .position(|window| window == needle)
+}
+
+pub struct ExtractedFile {
+    pub name: Option<String>,
+    pub data: Vec<u8>,
+}
+
+#[derive(Default)]
+pub struct OneDump<'a> {
+    embedded_files: Vec<ExtractedFile>,
+    remaining_vec: Option<Vec<u8>>,
+    remaining: Option<&'a [u8]>,
+}
+
+// https://learn.microsoft.com/en-us/openspecs/office_file_formats/ms-onestore/8806fd18-6735-4874-b111-227b83eaac26
+#[repr(packed)]
+struct FileDataHeader {
+    _guid_header: [u8; 16],
+    _cb_length: u64,
+    _unused: u32,
+    _reserved: u64,
+}
+
+impl<'a> OneDump<'a> {
+    pub fn from_bytes(data: &'a [u8], filename: &str) -> Result<OneDump<'a>, OneDumpError> {
+        debug!(
+            "Inspecting OneNote file for attachments from in-memory buffer of size {}-bytes named {}\n",
```

Comment:
It turns out I can't make this change if changing `filename` from `&str` to `&Path` because `Path` does not implement `Display`.  So I'll have to do `filename.to_string_loss()`.